### PR TITLE
flow:iio: Add Proximity sensor Category support into iio node

### PR DESCRIPTION
**Pull request by yongli3**
Signed-off-by: Yong Li <yong.b.li@intel.com>

### CODE REVIEW

**Comment by yongli3** [comment]
@lblim  @barbieri @edersondisouza please help to review, this PR is used for issue #1971

File: src/modules/flow/iio/iio.json [hunk 1]
```diff
@@ -803,6 +803,95 @@
         }
       ],
       "private_data_type": "light_data"
+     },
+	 {
```
Comment (edersondisouza) [inline src/modules/flow/iio/iio.json]: This `{` looks mis-indented

File: src/modules/flow/iio/iio.json [hunk 2]
```diff
@@ -803,6 +803,95 @@
...
+             "description": "Range of output packet. Usually, is the output range of used sensor. It'll be the 'min' and 'max' fields of drange packets sent on OUT port.",
+             "name": "out_range"
+         }
+       ],
+       "version": 1
+      },
+      "out_ports": [
+        {
+         "data_type": "float",
+         "description": "proximity data read in lux.",
```
Comment (edersondisouza) [inline src/modules/flow/iio/iio.json]: Proximity in lux?

File: src/modules/flow/iio/nodes.c [hunk 3]
```diff
@@ -1429,4 +1429,156 @@ light_tick(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_
...
+        .max = mdata->out_range.max,
+        .step = mdata->out_range.step
+    };
+    bool b;
+
+    b = sol_iio_read_channel_value(mdata->channel_val, &out.val);
+    if (!b) goto error;
+
+    sol_flow_send_drange_value_packet(node,
+        SOL_FLOW_NODE_TYPE_IIO_LIGHT__OUT__OUT, out.val);
```
Comment (edersondisouza) [inline src/modules/flow/iio/nodes.c]: `SOL_FLOW_NODE_TYPE_IIO_LIGHT__OUT__OUT`? Probably you wanted proximity node port...

**Comment by edersondisouza** [comment]
About the in_proximityX... how are they used? What does mean the data they read? A simple answer would be adding another option to node type, but I'd like to understand more about meaning of this data. Are they complementary? Do they measure different ranges?

**Comment by yongli3** [comment]
Thanks @edersondisouza  for the review. About the in_proximity data reading, for common sensors such as APDS9960/APDS9930, users will use "cat in_proximity_raw" to get the data, this PR can support such  kind of sensors. But if there are more than one raw data, such as SX9500 in issue #1971, we need to get data from in_proximity2_raw. @lblim could you share some information about these in_proximityX? do we need to read all of them, or just in_proximity2_raw?

File: src/modules/flow/iio/iio.json (hunk 1 above)
Comment (yongli3) [inline src/modules/flow/iio/iio.json]: Done

File: src/modules/flow/iio/iio.json (hunk 2 above)
Comment (yongli3) [inline src/modules/flow/iio/iio.json]: The data is raw data

File: src/modules/flow/iio/nodes.c (hunk 3 above)
Comment (yongli3) [inline src/modules/flow/iio/nodes.c]: Done

**Comment by yongli3** [comment]
@edersondisouza  @laykuanloon I modified this PR, it can support the default in_proximity_raw and in_proximity2_raw. 

If you want to support both these 4 channels, I think we can use another PR to implement it. @edersondisouza how do you think about it?

File: src/modules/flow/iio/nodes.c [hunk 4]
```diff
@@ -1429,4 +1429,158 @@ light_tick(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_
...
+    SOL_NULL_CHECK(mdata->device, false);
+
+#define ADD_CHANNEL(_val) \
+    if (!mdata->use_device_default_scale) \
+        channel_config.scale = mdata->scale; \
+    if (!mdata->use_device_default_offset) \
+        channel_config.offset = mdata->offset; \
+    mdata->channel_ ## _val = sol_iio_add_channel(mdata->device, "in_proximity", &channel_config); \
+    if (!mdata->channel_ ## _val) \
+        mdata->channel_ ## _val = sol_iio_add_channel(mdata->device, "in_proximity2", &channel_config); \
```
Comment (edersondisouza) [inline src/modules/flow/iio/nodes.c]: You are adding two channels, but only saves one. So, you only read `in_proximity2` in fact, as it's the last one you saved. If a device happens to not have `in_proximity2`, you'll get NULL here and no readings after.
Nevertheless, I still would like to understand the differences between these channels. I suspect that we could model all channels on only one final information.

Comment (edersondisouza) [inline src/modules/flow/iio/nodes.c]: Oh... disregard first comment, only now I've seen the `if (!mdata->channel_ ## _val)`... sorry.
But I'm kinda puzzled by the `in_proximity2`... AFAICU from sx9500 datasheet, there's no differentiation among channels...

Comment (yongli3) [inline src/modules/flow/iio/nodes.c]: @edersondisouza As discussed in issue #1971, only channel 2 in sx9500 is needed. 
APDS9960/APDS9930 only have the in_proximity_raw, and the SX9500 has in_proximity0_raw to in_proximity3_raw, if we only use in_proximity2_raw, this PR can support all these three sensors. But if we want to add multi-channel support, we need a new implement

Comment (edersondisouza) [inline src/modules/flow/iio/nodes.c]: Ok. I'll merge this PR, but we need to change nodes to address this multiple channel case.

**Comment by edersondisouza** [comment]
Integrated. Tks!
@yongli3 I'll create a Github issue about multiple channels and assign it for you, ok?

**Comment by yongli3** [comment]
No problem☺

**Pull request closed without merging**